### PR TITLE
Fix issue 20063 - Patch closure variables when converting expressions to delegates

### PR DESCRIPTION
**Pull request by SSoulaimane**
Issue [20063](https://issues.dlang.org/show_bug.cgi?id=20063).

### CODE REVIEW

File: src/dmd/delegatize.d [hunk 1]
```diff
@@ -86,6 +86,26 @@ private void lambdaSetParent(Expression e, FuncDeclaration fd)
         alias visit = typeof(super).visit;
         FuncDeclaration fd;
 
+        private void setParent(Dsymbol s)
+        {
+            if (s.isVarDeclaration() && s.parent.isFuncDeclaration())
```
Comment (Basile-z) [inline src/dmd/delegatize.d]: safer access and prevent double virtual calls with

```d
VarDeclaration vd = s.isVarDeclaration();
FuncDeclaration fd = s.parent ? s.parent.isFuncDeclaration() : null;
if (fd && vd) 
{  /*etc*/
```

File: src/dmd/delegatize.d [hunk 2]
```diff
@@ -86,6 +86,26 @@ private void lambdaSetParent(Expression e, FuncDeclaration fd)
         alias visit = typeof(super).visit;
         FuncDeclaration fd;
 
+        private void setParent(Dsymbol s)
+        {
+            VarDeclaration vd = s.isVarDeclaration();
+            FuncDeclaration pfd = s.parent ? s.parent.isFuncDeclaration() : null;            if (s.isVarDeclaration() && s.parent.isFuncDeclaration())
```
Comment (Basile-z) [inline src/dmd/delegatize.d]: little problem with the editor commands I suppose ? (check the end of this line)

Comment (SSoulaimane) [reply src/dmd/delegatize.d]: Ah, sorry. Thanks.

**Review by ghost — APPROVED** [review]

File: src/dmd/delegatize.d [hunk 3]
```diff
@@ -86,6 +86,26 @@ private void lambdaSetParent(Expression e, FuncDeclaration fd)
         alias visit = typeof(super).visit;
         FuncDeclaration fd;
 
+        private void setParent(Dsymbol s)
+        {
+            VarDeclaration vd = s.isVarDeclaration();
+            FuncDeclaration pfd = s.parent ? s.parent.isFuncDeclaration() : null;
+            if (vd && pfd)
```
Comment (thewilsonator) [inline src/dmd/delegatize.d]: please write this as 
```
s.parent = fd;
if (!vd || !pfd)
    return;
foreach (i; 0 .. pfd.closureVars.dim)
 {
      if (vd != pfd.closureVars[i])
          continue;

      pfd.closureVars.remove(i);
      fd.closureVars.push(vd);
      return;
}
```

**Review by thewilsonator — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/delegatize.d [hunk 4]
```diff
@@ -86,6 +86,25 @@ private void lambdaSetParent(Expression e, FuncDeclaration fd)
...
 
+        private void setParent(Dsymbol s)
+        {
+            VarDeclaration vd = s.isVarDeclaration();
+            FuncDeclaration pfd = s.parent ? s.parent.isFuncDeclaration() : null;
+            s.parent = fd;
+            if (!vd || !pfd)
+                return;
+            // move to fd's closure when applicable
+            foreach (i; 0 .. pfd.closureVars.dim)
```
Comment (ZombineDev) [inline src/dmd/delegatize.d]: BTW, one should be careful when using `foreach (; 0 .. max)` and changing the `max` as the emitted code caches it:
https://run.dlang.io/gist/run-dlang/b0b40d58474fb7c35b22803b2432c02f?compiler=dmd

Comment (SSoulaimane) [reply src/dmd/delegatize.d]: Yes, I notice that, but the code here breaks immediately after `remove()` is called.

**Pull request merged**